rewrite-core/src/main/java/org/openrewrite/table/RepositoryResults.java

```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Value
+    @With
+    public static class Row {
+        @Column(displayName = "Root of the recipe that made changes",
+                description = "The root recipe that made a change.")
+        String rootRecipe;
+
+        @Column(displayName = "First child recipe that made changes",
+                description = "The first child recipe that made a change.")
+        String recipe;
+
+        @Column(displayName = "Recipe options",
+                description = "The options of the first child (or root, if they are equal) recipe that made a change.")
+        String recipeOptions;
+
+        @Column(displayName = "Estimated time saving",
+                description = "An estimated effort that a developer to fix manually instead of using this recipe," +
+                              " in unit of seconds.")
+        Long estimatedTimeSaving;
+
+        @Column(displayName = "Occurrences",
+                description = "The total number of changes made to this repository.")
+        int occurrences;
+
+        @Column(displayName = "Files changed",
+                description = "The number of files changed to this repository.")
+        int filesChanged;
+    }
+
+    public void recordResult(@Nullable SourceFile before, Stack<Recipe> recipeStack, Long effortSeconds) {
+        int rootIndex = 0;
+
+        Recipe root = recipeStack.get(rootIndex);
+        // Skip any CompositeRecipe wrappers
+        while (root instanceof CompositeRecipe) {
+            root = recipeStack.get(++rootIndex);
+        }
+        Recipe recipe = recipeStack.size() < rootIndex + 1 ? root : recipeStack.get(rootIndex + 1);
+        String rootName = root == null ? "" : root.getName();
+        String recipeName = recipe == null ? "" : recipe.getName();
+        String options = recipe == null ? "[]" : recipe.getDescriptor().getOptions().stream()
+                .filter(option -> option.getValue() != null)
+                .map(option -> serializeOption(option.getName(), option.getValue()))
+                .collect(Collectors.joining(", ", "[", "]"));
+
+        Counter counter = counters.computeIfAbsent(new RecipeKey(rootName, recipeName, options), k -> new Counter());
+        if (before == null || counter.beforeFiles.add(before.getSourcePath())) {
+            counter.filesChanged++;
+        }
+        counter.occurrences++;
+        counter.estimatedTimeSavings += effortSeconds;
+    }
+
+    public void flush(ExecutionContext ctx) {
+        maxCycle = ctx.getCycle(); //allow to write to this data table after last cycle
+        for (Map.Entry<RecipeKey, Counter> entry : counters.entrySet()) {
+            insertRow(ctx, new Row(entry.getKey().getRootName(), entry.getKey().getRecipeName(), entry.getKey().getOptions(), entry.getValue().estimatedTimeSavings, entry.getValue().occurrences, entry.getValue().filesChanged));
+        }
+    }
+
+    public static String serializeOption(String name, Object value) {
```

Comment:
doubting between this more readable serialization or json (makes matching a lot easier)

Reply:
Wouldn't JSON basically be the same? Now you would get something like:
```
[foo="false", bar="1", baz="null", qux=`Foo"bar"`]
```
with JSON:
```json
{"foo":false, "bar":1, "baz":null, "qux":"Foo\"bar\""}
```
Of these I think JSON looks better and it also preserves basic data types and can again be deserialized as JSON.

Of course there will be some additional escaping when serialized as CSV. But that applies to both formats.